core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionImpl.java

```diff
@@ -78,23 +88,67 @@
                 parameters[i] = ctx;
             } else if (entry.getType().isAssignableFrom(HttpRequest.class)) {
                 parameters[i] = req;
+            } else if (entry.getType().isAssignableFrom(AggregatedHttpMessage.class)) {
+                parameters[i] = message;
             }
         }
         return parameters;
     }
 
     @Override
     public Object serve(ServiceRequestContext ctx, HttpRequest req, Map<String, String> args) throws Exception {
-        Object[] parameters = parameters(ctx, req, args);
+        if (aggregationRequired) {
+            if (CompletionStage.class.isAssignableFrom(method.getReturnType())) {
+                return req.aggregate()
+                          .thenComposeAsync(msg -> (CompletionStage<?>) invokeMethod(ctx, req, args, msg),
+                                            ctx.contextAwareEventLoop());
+            } else if (isAsynchronous) {
+                return req.aggregate()
+                          .thenApplyAsync(msg -> invokeMethod(ctx, req, args, msg),
+                                          ctx.contextAwareEventLoop());
```

Comment:
How about just using `thenApply()` and `thenCompose()` instead of `*Async()`? Is there any reason to use it?